components/infobox/extensions/commons/age_calculation.lua

```diff
@@ -0,0 +1,154 @@
+local _CURRENT_YEAR = tonumber(_LANG:formatDate('Y'))
+local _DAYS_IN_MONTH = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 }
+
+function AgeCalculation.run(birth, birthLocation, death, personType, shouldStore)
+	--disect the dates into fields
+	local birthFields = mw.text.split(birth or '', '-')
+	local deathFields = mw.text.split(death or '', '-')
+	--strip them to numbers
+	for i = 1, 3 do
+		birthFields[i] = tonumber(birthFields[i] or '')
+		deathFields[i] = tonumber(deathFields[i] or '')
+	end
+
+	birthFields = AgeCalculation._processDateFields(birthFields)
+	deathFields = AgeCalculation._processDateFields(deathFields)
+	AgeCalculation._sanityCeck(birthFields, deathFields)
+	local age = AgeCalculation._processAge(birthFields, deathFields)
+
+	local birthDisplay
+	if birthFields.display then
+		birthDisplay = birthFields.display .. age.birth
+		if birthLocation then
+			birthDisplay = birthDisplay .. '<br>' .. birthLocation
+		end
+		if shouldStore and not birthFields.exact then
+			birthDisplay = birthDisplay .. '[[Category:Incomplete birth dates]]'
+		end
+		if shouldStore and birthFields[1] then
+			birthDisplay = birthDisplay .. '[[Category:' .. birthFields[1] .. 'births]]'
+		end
+	end
+	local deathDisplay
+	if deathFields.display then
+		deathDisplay = deathFields.display .. age.death
+		if shouldStore and not deathFields.exact then
+			deathDisplay = deathDisplay .. '[[Category:Incomplete death dates]]'
+		end
+	end
+
+	return birthDisplay, deathDisplay, birthFields.store, deathFields.store
+end
+
+function AgeCalculation._processAge(birthFields, deathFields)
+	local age = { birth = '', death = '' }
+	if not birthFields.display then
+		return age
+	end
+	if not deathFields.display then
+		deathFields[1] = _CURRENT_FIELD['year']
+		deathFields[2] = _CURRENT_FIELD['month']
+		deathFields[3] = _CURRENT_FIELD['day']
+		deathFields.exact = true
+	end
+	local calculatedAgeDisplay
+	if birthFields.exact and deathFields.exact then
+		calculatedAgeDisplay = AgeCalculation._calculateAge(birthFields, deathFields)
+	elseif deathFields[1] and birthFields[1] then
+		local minAge = AgeCalculation._calculateAge(birthFields.max or birthFields, deathFields.min or deathFields)
+		local maxAge = AgeCalculation._calculateAge(birthFields.min or birthFields, deathFields.max or deathFields)
+		if minAge == maxAge then
+			calculatedAgeDisplay = minAge
+		else
+			calculatedAgeDisplay = minAge .. '-' .. maxAge
+		end
+	end
+
+	if deathFields.display and calculatedAgeDisplay then
+		age.death = ' (aged ' .. calculatedAgeDisplay .. ')'
+	elseif calculatedAgeDisplay then
+		age.birth = ' (age ' .. calculatedAgeDisplay .. ')'
+	end
+
+	return age
+end
+
+function AgeCalculation._calculateAge(startDate, endDate)
+	local age = endDate[1] - startDate[1]
+	if
+		endDate[2] < startDate[2] or (
+			endDate[2] == startDate[2] and endDate[3] < startDate[3]
+		)
+	then
+		age = age - 1
+	end
+	return age
+end
+
+function AgeCalculation._processDateFields(date)
```

Comment:
This function sounds like it's just doing some simple processing, I thought this was just turning the numerical values into dates. Why is it doing so much? I'm basically lost from this function on.

Reply:
added "some" comments to describe what is happening

Reply:
The comments help, but I don't understand why we're making this so complicated. Why are year, month and day equal to `1, 2, 3`, and not just keys with the names `year, month, day`?  What is min? What is max? And most importantly, and this is what I was trying to say before, why is this all in one function? I don't see how a function consisting of 50 lines of processing could ever be the right answer.

Can't we just do it one by one? Now we end up with a flow where, as a reader, you're supposed to know that `birthFields[1]` apparently equals the year. You basically can't understand what's going on without reading all of the code, which goes against the point of abstraction.

Reply:
adjusted it
* broke down function `_processDateFields` into several parts
* rename date keys from `1`, `2`, `3` to `year`, `month`, `day`